internal/controllers/finalizers/specialresource.go

```diff
@@ -0,0 +1,166 @@
+package finalizers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+	"github.com/openshift-psap/special-resource-operator/api/v1beta1"
+	"github.com/openshift-psap/special-resource-operator/pkg/cache"
+	"github.com/openshift-psap/special-resource-operator/pkg/clients"
+	"github.com/openshift-psap/special-resource-operator/pkg/poll"
+	"github.com/openshift-psap/special-resource-operator/pkg/state"
+	"github.com/openshift-psap/special-resource-operator/pkg/utils"
+	"github.com/pkg/errors"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const FinalizerString = "sro.openshift.io/finalizer"
+
+type SpecialResourceFinalizer interface {
+	AddToSpecialResource(ctx context.Context, sr *v1beta1.SpecialResource) error
+	Finalize(ctx context.Context, sr *v1beta1.SpecialResource) error
+}
+
+type specialResourceFinalizer struct {
+	kubeClient  clients.ClientsInterface
+	log         logr.Logger
+	nodesCacher cache.NodesCacher
+	pollActions poll.PollActions
+}
+
+func NewSpecialResourceFinalizer(
+	kubeClient clients.ClientsInterface,
+	nodesCacher cache.NodesCacher,
+	pollActions poll.PollActions,
+) SpecialResourceFinalizer {
+	return &specialResourceFinalizer{
+		kubeClient:  kubeClient,
+		log:         ctrl.Log.WithName("finalizers"),
+		nodesCacher: nodesCacher,
+		pollActions: pollActions,
+	}
+}
+
+func (srf *specialResourceFinalizer) AddToSpecialResource(ctx context.Context, sr *v1beta1.SpecialResource) error {
+	srf.log.Info("Adding finalizer to special resource")
+	controllerutil.AddFinalizer(sr, FinalizerString)
+
+	// Update CR
+	if err := srf.kubeClient.Update(ctx, sr); err != nil {
+		srf.log.Info("Adding finalizer failed", "error", fmt.Sprintf("%v", err))
```

Comment:
Nit: I see this code only changed place, but maybe we could use `srf.log.Error(err, "msg")` instead?

Reply:
Great. No problem with doing it in another PR